### PR TITLE
Benchmark and speed processing of polyglot java imports up

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Introduces new startup "Hello World" [benchmark with many imports](https://github.com/enso-org/enso/pull/10899/commits/f4a29a1c06aaa5b7d5699b78a4d1cdfef1273a17) and a performance improvement by [delaying loading classes in registerPolyglotSymbol](https://github.com/enso-org/enso/pull/10899/commits/f93ee2166b258d031d75bb1d3af1a18145d0be98).

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- [x] Benchmarks results are reasonable
    - [x] Initial [stdlib benchmarks run](https://github.com/enso-org/enso/actions/runs/10596232710) - confirms there is [22% slowdown due to additional imports](https://github.com/enso-org/enso/pull/10899#discussion_r1734822871)
    - [x] Final [stdlib benchmarks run](https://github.com/enso-org/enso/actions/runs/10596279560)

Original motivation for this PR is @hubertp `profilling.npss` that spends **700ms** in registering polyglot symbols (which means to load all the Java classes referenced by `polyglot import java` statements and classloading takes time):

2s

Hubert remeasured the behavior on his slow computer and here is the result:

1s

The `registerPolyglotImports` method is completely gone and the **whole** `processModule` takes just 1.1s (previously 1.9s). That seems like an improvement we want to integrate.

### CODE REVIEW

**Review by jdunkerley — APPROVED** [review]

**Comment by JaroslavTulach** [comment]
Attempting to run stdlib benchmarks with f93ee21: [first attempt](https://github.com/enso-org/enso/actions/runs/10580026830).

**Review by GregoryTravis — APPROVED** [review]

**Comment by JaroslavTulach** [comment]
We have two commits right now:
- [Benchmark measuring startup with import of all libraries](https://github.com/enso-org/enso/pull/10899/commits/f4a29a1c06aaa5b7d5699b78a4d1cdfef1273a17) - f4a29a1c06aaa5b7d5699b78a4d1cdfef1273a17
- [Delay loading classes in registerPolyglotSymbol by using Suppliers](https://github.com/enso-org/enso/pull/10899/commits/f93ee2166b258d031d75bb1d3af1a18145d0be98) - f93ee2166b258d031d75bb1d3af1a18145d0be98

now we need to restructure the PR and do new benchmark measurements because of 
- #10902 

183ca7a should contain all the necessary fixes as well as [Benchmark measuring startup with import of all libraries](https://github.com/enso-org/enso/pull/10899/commits/f4a29a1c06aaa5b7d5699b78a4d1cdfef1273a17). Benchmark is running at https://github.com/enso-org/enso/actions/runs/10596232710

44ca279 is a follow up of 183ca7a and contains also fix for [Delay loading classes in registerPolyglotSymbol by using Suppliers](https://github.com/enso-org/enso/pull/10899/commits/f93ee2166b258d031d75bb1d3af1a18145d0be98). Benchmark will run at https://github.com/enso-org/enso/actions/runs/10596279560

File: test/Benchmarks/src/Startup/Import_World.enso [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from Standard.Base import all
```
Comment (JaroslavTulach) [inline test/Benchmarks/src/Startup/Import_World.enso:1]: This example was obtained by running `enso --new HelloExample` and then the body of `main` was changed.

**Comment by JaroslavTulach** [comment]
There is an [ExecCompilerTest](https://github.com/enso-org/enso/actions/runs/10596264181/job/29363853514?pr=10899#step:7:5438) failure caused by running with `micro-distribution`. Fixed by 0fba66b

File: test/Benchmarks/src/Startup/Import_World.enso (hunk 1 above)
Comment (JaroslavTulach) [reply test/Benchmarks/src/Startup/Import_World.enso:1]: Startup with **all imports** is **22% slower** than startup with simple `from Standard.Base import IO`:
- this is the [benchmark run](https://github.com/enso-org/enso/actions/runs/10596232710)
- `org_enso_benchmarks_generated_Startup_hello_world_startup` is **3453ms**
- `org_enso_benchmarks_generated_Startup_import_world_startup` is **4217ms**

E.g. a _typical empty IDE project_ wastes 22% processing modules that aren't needed for the execution.

Comment (JaroslavTulach) [reply test/Benchmarks/src/Startup/Import_World.enso:1]: The second [stdlib benchmarks run](https://github.com/enso-org/enso/actions/runs/10596279560) has improved all the startup benchmarks, but only minimally by 4%%, 1%% and 17%% - that's below the measurement precision we have. I was expecting bigger speedup according to the VisualVM graphs.

Expected 700ms

Comment (JaroslavTulach) [reply test/Benchmarks/src/Startup/Import_World.enso:1]: The new version 0fba66b removes any trace of `registerPolyglotSymbol` from the (10ms interval) sampling. That's a good sign. However there still seems to be `getPolyglotSymbol` caught in the trace two times:

getPolyglotSymbol

Maybe we could improving the startup time by using `Supplier<Object>` where the _polyglot symbol_ is requested by the `IrToTruffle`. One more [benchmarks run](https://github.com/enso-org/enso/actions/runs/10603589951) scheduled with https://github.com/enso-org/enso/commit/6ee9c8f80feb3caa8ec26b2d735451efd22c91cd

Comment (JaroslavTulach) [reply test/Benchmarks/src/Startup/Import_World.enso:1]: Previous benchmark run failed. Scheduling [new one](https://github.com/enso-org/enso/actions/runs/10615005609) for e2c7bd3

**Comment by JaroslavTulach** [comment]
The introduction of _error message_ for `DataflowError` (done in 6c00db4) causes a [WarningsTest failure](https://github.com/enso-org/enso/actions/runs/10611440276/job/29410971157?pr=10899#step:7:4348) caused by the fact that now there is `ex.getMessage() != null` for `DataflowError`. That's good, but as warnings cannot be attached to `DataflowError`s - we need a change in the test. Done in 53eece8f3a30231a9f095714ec2b3120b517a2ba

File: engine/runtime-integration-tests/src/test/java/org/enso/compiler/ExecCompilerTest.java [hunk 2]
```diff
@@ -408,16 +408,15 @@ public void testDoubledRandom() throws Exception {
...
           run seed =
-              operator1 = Random.new_generator seed
+              Random.new_generator seed
           """);
     var run = module.invokeMember("eval_expression", "run");
     try {
       var err = run.execute(1L);
-      fail("Not expecting any result: " + err);
+      assertTrue("Returned value represents exception: " + err, err.isException());
+      throw err.throwException();
```
Comment (JaroslavTulach) [inline engine/runtime-integration-tests/src/test/java/org/enso/compiler/ExecCompilerTest.java:417]: Resolving `polyglot java import`s lazily prevents us to emit error during compilation. Instead we get a (dataflow) `Error` when we execute the `run` function and try to resolve `Random` function. Such an error is then returned as a value and needs to be `throwException` manually.

File: engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/WarningsTest.java [hunk 3]
```diff
@@ -148,16 +148,27 @@ private void assertWarningsForAType(Value v) {
...
+        try {
+          assertThat(
+              "Warning found for " + type,
+              ex.getMessage(),
+              AllOf.allOf(containsString("warn:once"), containsString("warn:twice")));
+        } catch (AssertionError err) {
+          if (type != null && v.equals(warning1) && v.equals(warning2)) {
+            assertEquals(
+                "Cannot attach warnings to Error - check it is an error",
+                "Standard.Base.Error.Error",
```
Comment (JaroslavTulach) [inline engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/WarningsTest.java:166]: Special case to handle [DataflowErrors that cannot have Warnings](https://github.com/enso-org/enso/pull/10899#issuecomment-2317082738).

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/constant/LazyObjectNode.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    return new LazyObjectNode(errorMessage, supplier);
+  }
+
+  @Override
+  public Object executeGeneric(VirtualFrame frame) {
+    var result = supply.get();
+    if (result instanceof TruffleObject) {
+      return result;
+    }
+    return DataflowError.withDefaultTrace(Text.create(error), this);
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/constant/LazyObjectNode.java:42]: You only check if it is a `TruffleObject` not if it is `null` which seems to be what javadoc suggests?

**Review by hubertp — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/constant/LazyObjectNode.java (hunk 4 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/constant/LazyObjectNode.java:42]: Do you mean the Javadoc of [build method](https://github.com/enso-org/enso/pull/10899/files/53eece8f3a30231a9f095714ec2b3120b517a2ba#diff-7811c280da1db3c47bf9c385c5d9453840df54721b64d672425f1745e7ddc28bR32) that takes `Supplier<TruffleObject>` argument?

If typing is on, then check for `null` is the same as `instanceof TruffleObject`. But I decided to go for the latter as someone could sneak in raw `Supplier` and return anything. I would need to perform some complicated checks to verify the value is proper _Enso interpreter value_. That can be added, but for now `instanceof TruffleObject` is sufficient.

**Comment by JaroslavTulach** [comment]
At the end this PR has managed to speed the new `Import_Hello` benchmark by 4% from 4217ms to 4063ms:

4% in Import Hello

**Pull request merged**